Add check and remove temporary fix for lg soundbar eq and source list

bernimoses opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Adds a temporary fix for the eq and source list of lg soundbars since the used library does not handle unknown sources.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #32832
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bernimoses committed:
Add temporary fix for lg soundbar eq and source list

homeassistant commented:
Hi @bernimoses, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#38798).

Once you do that we will be able to review and accept this pull request.

Thanks!

frenck commented on homeassistant/components/lg_soundbar/media_player.py:
```diff
@@ -130,6 +130,11 @@ def update(self):
             if equaliser >= len(temescal.equalisers):
                 temescal.equalisers.append("unknown " + str(equaliser))
 
+        # Temporary fix until handling of unknown function settings is integrated in the temescal library
+        for function in self._functions:
+            if function >= len(temescal.functions):
+                temescal.functions.append("unknown " + str(function))
+
```
There is nothing more permanent than a temporary fix.

We should not add this to Home Assistant, instead the upstream library needs to be fixed first.

bernimoses replied:
Ok, i will make a pull request for the original library. Would it be ok to keep this pull request (and later merge it) in case the library author does not respond in the next couple of weeks?

akinomeroglu commented:
Hi,

I just reached temascal maintainer and he said that he will release a new version which will support new eq settings soon. I believe that this problem will solved when he released the new version and then our version bump.

bernimoses commented:
Sounds good to me if the new support is not just adding new variables (like my pull request i sent him). Since a new soundbar with new options would break it again.

I still have some other problems with the integration / library like keeping a long-lived connection without polling since that wakes the soundbar every time. Maybe i can get in touch with the maintainer too to tackle some of them.

akinomeroglu commented:
@bernimoses Hi, library maintainer released 0.2 yesterday. Unfortunately, it does not cover your pr as far as I see. I created a pr against core for version bump #39213 

Thanks and regards,

bernimoses commented:
@akinomeroglu Thank you. There is already a new version 0.3 with some bugfixes. This does not cover all future eq and source list types but at least the current ones i am aware of are taken care of. Do you want to create a version bump for this or should i try it. ;) Thanks

akinomeroglu commented:
Hi @bernimoses,

I will be my honor if you can bump it this time. =) Thanks for your effort on this subject. If I understand correctly we can close this PR after version update?

Regards,

bernimoses commented:
@akinomeroglu finally had time to create the version bump #39379.

Not sure if we should close this, because it just fixes my problem for now but newer versions of a soundbar could have the same problem again. It should be possible to remove the 2 temporary fix blocks within `handle_event`. This way there should be no python error anymore (because of the array length check). The downside is there is no way to select unknown `equalisers` or `functions` the library does not define. What do you think?

bernimoses commented:
Updated the patch to remove the "temporary fix" and added a array check for the source list (same as eq list). Using it for a month and had no problems/exceptions.

bernimoses commented:
I think there should be a check if the array index exists in line [163](https://github.com/home-assistant/core/blob/9893d30485a987a617e5c4783e9aeca2c7e11c71/homeassistant/components/lg_soundbar/media_player.py#L163) and [178](https://github.com/home-assistant/core/blob/9893d30485a987a617e5c4783e9aeca2c7e11c71/homeassistant/components/lg_soundbar/media_player.py#L178) to get this finally fixed (avoid errors like [this](https://github.com/home-assistant/core/issues/41121#issuecomment-720333642)).

@frenck is this something you would consider merging if i added it or is there something else i can do?

frenck reviewed (approved):
This looks good now! 

Thanks, @bernimoses 👍

bernimoses commented:
@frenck not sure if this is completely fixed (last comment).

> I think there should be a check if the array index exists in line [163](https://github.com/home-assistant/core/blob/9893d30485a987a617e5c4783e9aeca2c7e11c71/homeassistant/components/lg_soundbar/media_player.py#L163) and [178](https://github.com/home-assistant/core/blob/9893d30485a987a617e5c4783e9aeca2c7e11c71/homeassistant/components/lg_soundbar/media_player.py#L178) to get this finally fixed (avoid errors like [this](https://github.com/home-assistant/core/issues/41121#issuecomment-720333642)).

frenck commented:
@bernimoses My impression was that that was optional. But you can still do that in a follow-up PR if you consider that needed.
